ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java

```diff
@@ -435,64 +434,86 @@ public void handleDeleteRequest(NetworkRequest request) throws IOException, Inte
+    DeleteResponse response;
+    MessageInfo info = null;
+    StoreKey convertedStoreKey = null;
+    Store storeToDelete = null;
+    ServerErrorCode serverErrorCode = ServerErrorCode.No_Error;
     try {
-      StoreKey convertedStoreKey = getConvertedStoreKeys(Collections.singletonList(deleteRequest.getBlobId())).get(0);
-      ServerErrorCode error =
+      convertedStoreKey = getConvertedStoreKeys(Collections.singletonList(deleteRequest.getBlobId())).get(0);
+      serverErrorCode =
           validateRequest(deleteRequest.getBlobId().getPartition(), RequestOrResponseType.DeleteRequest, false);
-      if (error != ServerErrorCode.No_Error) {
-        logger.error("Validating delete request failed with error {} for request {}", error, deleteRequest);
-        response = new DeleteResponse(deleteRequest.getCorrelationId(), deleteRequest.getClientId(), error);
+      if (serverErrorCode != ServerErrorCode.No_Error) {
+        logger.error("Validating delete request failed with error {} for request {}", serverErrorCode, deleteRequest);
       } else {
         BlobId convertedBlobId = (BlobId) convertedStoreKey;
-        MessageInfo info = new MessageInfo.Builder(convertedBlobId, -1, convertedBlobId.getAccountId(),
+        info = new MessageInfo.Builder(convertedBlobId, -1, convertedBlobId.getAccountId(),
             convertedBlobId.getContainerId(), deleteRequest.getDeletionTimeInMs()).isDeleted(true)
             .lifeVersion(MessageInfo.LIFE_VERSION_FROM_FRONTEND)
             .build();
-        Store storeToDelete = storeManager.getStore(deleteRequest.getBlobId().getPartition());
+        storeToDelete = storeManager.getStore(deleteRequest.getBlobId().getPartition());
         storeToDelete.delete(Collections.singletonList(info));
-        response =
-            new DeleteResponse(deleteRequest.getCorrelationId(), deleteRequest.getClientId(), ServerErrorCode.No_Error);
-        if (notification != null) {
-          notification.onBlobReplicaDeleted(currentNode.getHostname(), currentNode.getPort(), convertedStoreKey.getID(),
-              BlobReplicaSourceType.PRIMARY);
-        }
       }
     } catch (StoreException e) {
-      boolean logInErrorLevel = false;
-      if (e.getErrorCode() == StoreErrorCodes.ID_Not_Found) {
-        metrics.idNotFoundError.inc();
-      } else if (e.getErrorCode() == StoreErrorCodes.TTL_Expired) {
-        metrics.ttlExpiredError.inc();
-      } else if (e.getErrorCode() == StoreErrorCodes.ID_Deleted) {
-        metrics.idDeletedError.inc();
-      } else if (e.getErrorCode() == StoreErrorCodes.Authorization_Failure) {
-        metrics.deleteAuthorizationFailure.inc();
+      if (e.getErrorCode() == StoreErrorCodes.ID_Not_Found && deleteRequest.shouldForceDelete()) {
+        try {
+          // If frontend forces a delete operation, place a tombstone even though blob is not present
+          Objects.requireNonNull(storeToDelete).forceDelete(Collections.singletonList(info));
```

Comment:
if we are here, the `storeToDelete` will not be null.

Reply:
Yeah.. true.